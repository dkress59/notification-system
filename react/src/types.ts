import { DetailedHTMLProps, HTMLAttributes, PropsWithChildren } from 'react'

import { SpawnBannerProps, SpawnModalProps, SpawnToastProps } from ':core/types'

type CustomHTMLAttributes<
	T,
	K = void | Record<string, unknown>,
> = DetailedHTMLProps<HTMLAttributes<T> & K, T>

export type BannerAreaAttributes = CustomHTMLAttributes<HTMLBannerAreaElement>

export type NotificationAreaAttributes =
	CustomHTMLAttributes<HTMLNotificationAreaElement>

export type BannerNotificationAttributes =
	CustomHTMLAttributes<HTMLBannerNotificationElement>

export type ModalNotificationAttributes =
	CustomHTMLAttributes<HTMLModalNotificationElement>

export type ToastNotificationAttributes =
	CustomHTMLAttributes<HTMLToastNotificationElement>

declare global {
	// eslint-disable-next-line @typescript-eslint/no-namespace
	namespace JSX {
		interface IntrinsicElements {
			['banner-area']: BannerAreaAttributes
			['notification-area']: NotificationAreaAttributes
			['banner-notification']: BannerNotificationAttributes
			['modal-notification']: ModalNotificationAttributes
			['toast-notification']: ToastNotificationAttributes
		}
	}
}

export type BannerAreaProps = PropsWithChildren

export type NotificationAreaProps = PropsWithChildren

export interface BannerNotificationProps
	extends PropsWithChildren,
		Omit<SpawnBannerProps, 'content'> {
	onDismiss?: () => void
}

export interface ToastNotificationProps
	extends PropsWithChildren,
		Omit<SpawnToastProps, 'content'> {
	onDismiss?: () => void
}

export interface ModalNotificationProps
	extends PropsWithChildren,
		Omit<SpawnModalProps, 'content'> {
	onConfirm?: () => void
	onDecline?: () => void
	onDismiss?: () => void
}
