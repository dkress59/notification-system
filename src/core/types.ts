import {
	HTMLBannerAreaElement,
	HTMLBannerNotificationElement,
	HTMLModalNotificationElement,
	HTMLNotificationAreaElement,
	HTMLToastNotificationElement,
} from './components'

type SnakeToCamelCaseProp<S extends string> = S extends
	| `${infer T} ${infer U}`
	| `${infer T}-${infer U}`
	? `${Uncapitalize<T>}${Capitalize<SnakeToCamelCaseProp<U>>}`
	: Uncapitalize<S>
type SnakeToCamelCase<T> = {
	[K in keyof T as SnakeToCamelCaseProp<K & string>]: T[K]
}

export enum NotificationType {
	SUCCESS = 'success',
	INFO = 'info',
	ERROR = 'error',
	WARNING = 'warning',
}

export enum NotificationEvent {
	BANNER_DISMISSED = 'bannerDismissed',
	MODAL_CONFIRMED = 'modalConfirmed',
	MODAL_DECLINED = 'modalDeclined',
	MODAL_DISMISSED = 'modalDismissed',
	TOAST_DISMISSED = 'toastDismissed',
}

type SpawnArgs<T> = T & {
	content: string
}

export interface BannerAttributes {
	'auto-hide'?: boolean | string
	'auto-hide-after-ms'?: number | string
	'headline'?: string
	'type'?: NotificationType
}
export type SpawnBannerArgs = SpawnArgs<SnakeToCamelCase<BannerAttributes>>

export interface ModalAttributes {
	'condition'?: boolean
	'headline'?: string
	'show-confirm'?: boolean
	'show-decline'?: boolean
	'type'?: NotificationType
	'label-confirm'?: string
	'label-decline'?: string
}
export type SpawnModalArgs = SpawnArgs<SnakeToCamelCase<ModalAttributes>>

export interface ToastAttributes {
	'auto-hide'?: boolean | string
	'auto-hide-after-ms'?: number | string
	'headline'?: string
	'type'?: NotificationType
}
export type SpawnToastArgs = SpawnArgs<SnakeToCamelCase<ToastAttributes>>

declare global {
	interface HTMLElementTagNameMap {
		'banner-area': HTMLBannerAreaElement
		'banner-notification': HTMLBannerNotificationElement
		'modal-notification': HTMLModalNotificationElement
		'notification-area': HTMLNotificationAreaElement
		'toast-notification': HTMLToastNotificationElement
	}
}
