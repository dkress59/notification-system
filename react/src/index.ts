export * from './components/banner-area'
export * from './components/banner-notification'
export * from './components/modal-notification'
export * from './components/notification-area'
export * from './components/toast-notification'
export * from './types'
