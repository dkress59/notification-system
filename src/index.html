<!DOCTYPE html>
<html dir="ltr" lang="en">

<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=5.0" />
  <title>Notification Web Components</title>

  <script type="module" src="/build/notification-system.esm.js"></script>
  <script nomodule src="/build/notification-system.js"></script>

  <style>
    * {
      box-sizing: border-box;
    }

    body {
      font-family: sans-serif;
    }

    #notification-container {
      position: fixed;
      z-index: 50;
      top: 0;
      right: 0;
      bottom: 0;
      left: 0;
      pointer-events: none;
      display: flex;
      flex-direction: column;
      align-items: flex-end;
      overflow-y: auto;
    }

    /* #notification-container > * {
        right: 0;
      } */
  </style>
  <script>
    document.addEventListener('DOMContentLoaded', () => {
      document.querySelectorAll('notify-toast').forEach(toast => {
        toast.addEventListener('toastDismissed', (shadowRoot) => console.debug(
          'toast dismissed!',
          { shadowRoot }
        ))
      })
    })
  </script>
</head>

<body>
  <div id="notification-container">
    <!-- <my-component first="Stencil" last="'Don't call me a framework' JS"></my-component>
      <my-component first="more Stencil" middle="please" last="'Don't call me a framework' JS"></my-component> -->
    <notify-toast type="error" headline="Lorem Ipsum">Lorem ipsum dolor sit amet consectetur adipisicing elit, facilis
      necessitatibus sunt quisquam officia quidem.</notify-toast>
    <notify-toast>Mock Content</notify-toast>
    <notify-toast auto-hide>Mock Content</notify-toast>
    <notify-toast type="info" is-hidden>Hidden Content</notify-toast>
    <notify-toast>Mock Content</notify-toast>
  </div>
</body>

</html>