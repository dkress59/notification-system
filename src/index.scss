@import 'css/base.scss';
// @import 'css/util.scss';
