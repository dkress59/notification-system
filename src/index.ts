export { Components, JSX } from './components'
export { BannerArea } from './components/banner-area/banner-area'
export { BannerNotification } from './components/banner-notification/banner-notification'
export { ModalNotification } from './components/modal-notification/modal-notification'
export { NotificationArea } from './components/notification-area/notification-area'
export { ToastNotification } from './components/toast-notification/toast-notification'
