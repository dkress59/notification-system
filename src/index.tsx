import './index.scss'

import React from 'react'
import ReactDOM from 'react-dom'

import { App } from './app'
import { NotificationProvider } from './hooks'

ReactDOM.render(
	<React.StrictMode>
		<NotificationProvider>
			<App />
		</NotificationProvider>
	</React.StrictMode>,
	document.getElementById('app'),
)
