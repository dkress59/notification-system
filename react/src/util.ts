import { ForwardedRef, MutableRefObject } from 'react'

export function getCurrentRef<T extends HTMLElement>({
	internalRef,
	ref,
}: {
	internalRef: MutableRefObject<null | T>
	ref: ForwardedRef<T>
}): T | null {
	if (ref) return (ref as MutableRefObject<null | T>).current
	return internalRef.current
}
