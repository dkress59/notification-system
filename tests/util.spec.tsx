import { act, render } from '@testing-library/react'

import { ToastType } from '../src/types'
import { getCollectionMapper } from '../src/util'

describe('utilities', () => {
	describe('getCollectionMapper', () => {
		it('correctly maps collection', () => {
			const mockRemoveFromDom = jest.fn()
			const mockMapper = getCollectionMapper(mockRemoveFromDom)
			expect(
				mockMapper({
					props: {
						autoHide: false,
						children: <>Mock Content</>,
						title: 'Mock Title',
						type: ToastType.INFO,
					},
					id: 'mock-id',
				}),
			).toMatchSnapshot()
		})
		it('passes in the removeToastFromDom function', async () => {
			const mockRemoveFromDom = jest.fn()
			const mockMapper = getCollectionMapper(mockRemoveFromDom)
			const mappedCollection = mockMapper({
				props: {
					autoHide: true,
					children: <>Mock Content</>,
					title: 'Mock Title',
					type: ToastType.INFO,
				},
				id: 'mock-id',
			})
			render(mappedCollection)
			await act(
				() => new Promise(resolve => setTimeout(() => resolve(), 3600)),
			)
			expect(mockRemoveFromDom).toHaveBeenCalled()
		})
	})
})
