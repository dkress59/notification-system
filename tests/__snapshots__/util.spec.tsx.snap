// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utilities getCollectionMapper correctly maps collection 1`] = `
<Toast
  autoHide={false}
  removeToastFromDom={[Function]}
  title="Mock Title"
  type="info"
>
  <React.Fragment>
    Mock Content
  </React.Fragment>
</Toast>
`;
